classes/converter.php

```diff
@@ -44,6 +45,26 @@ abstract class converter {
      */
     protected const NAME = 'base';
 
+    /**
+     * Generate the PDF content of a target URL passed through proxy URL.
+     *
+     * @param \moodle_url $proxyurl the plugin proxy url for access key login and redirection to target URL
+     * {@link /admin/tool/pages/index.php}.
+     * @param string $filename the name to give converted file.
+     * @param array $options any additional options to pass to converter, valid options vary with converter
+     * instance, see relevant converter for further details.
+     * @param string $cookiename cookie name to apply to conversion (optional).
+     * @param string $cookievalue cookie value to apply to conversion (optional).
+     *
+     * @return string raw PDF content of URL.
+     */
+    protected function generate_pdf_content(moodle_url $proxyurl, string $filename = '', array $options = [],
```

Comment:
this one should be abstract to make sure that children implement it.

Reply:
also should it be named as generate_pdf_file? as that what the method does.

Reply:
ah scratch it. It does generate a content and return a string.

Reply:
Abstracted the method